### ISSUE TITLE
replacement tx not detected as replacement (replacing: [])

### ISSUE THREAD

**Issue by kallerosenbaum** [dgarage/NBXplorer#421]
A `newtransaction` event contains a `replacing` property that lists all tracked transactions that this new transaction replaces.

This works fine most of the time, but at times NBX fails to set the replacing property correctly, and instead sets it to `[]`.

We have two NBX instances tracking the same set of addresses and both of them shows the same behavior on every address (I mean for a specific replacement, either both instances show the replacing property properly, or both fails to do it).

An example is txid `96e08ce4b137fd8025fcc7cbc529444e90deb094b65988d8311daed257a3663c` which is replaced by `85b52cecc22077c6a02cf7d45c0452d6f5333b1dbdc02e5ddcdf591da0c19f90`

The `newtransaction` events  stored in the NBX database for these transactions are, respectively:

```
Event ID: 531245 (Created at 2023-08-21 07:45:50.946705 +00:00)
{"data": {"blockId": null, "outputs": [{"index": 0, "value": 290435, "address": "bc1q8xv78yyuf3eqpch7w4f8xnalz3nvu48qzn5gzy", "keyPath": "0/YYYYY", "scriptPubKey": "00143999e3909c4c7200e2fe7552734fbf1466ce54e0"}], "replacing": [], "cryptoCode": "BTC", "trackedSource": "DERIVATIONSCHEME:xpubXXXXXXXXXXXXXXXXXXXX", "transactionData": {"height": null, "blockId": null, "timestamp": 1692603950, "replacedBy": null, "transaction": "02000000000101e3934c150e3b249fe1a7b80f319197e49eae4610826365a6a20b68e8154ea86f01000000000000008002836e0400000000001600143999e3909c4c7200e2fe7552734fbf1466ce54e0514825000000000016001439e9d8d06cc61f11ccff05c8a1afa172e44e6e4d0248304502210081e289c7ddcee3b3e72b6b41597e58dcfd72c780598aee34d93b927683989f3502203c9f8e4628eb34b47317ddabc5cf50801ec2fe999e02aa2e580457915206373f012103ef68e0004ec17595fd920aaee7872e01d3a6f0f55842d0fe58066b00484b89d100000000", "confirmations": 0, "transactionHash": "96e08ce4b137fd8025fcc7cbc529444e90deb094b65988d8311daed257a3663c"}, "derivationStrategy": "xpubXXXXXXXXXXXXXXXXXXXX"}, "type": "newtransaction"}

Event ID: 531247 (Created at 2023-08-21 07:46:33.605007 +00:00)
{"data": {"blockId": null, "outputs": [{"index": 0, "value": 290435, "address": "bc1q8xv78yyuf3eqpch7w4f8xnalz3nvu48qzn5gzy", "keyPath": "0/YYYYY", "scriptPubKey": "00143999e3909c4c7200e2fe7552734fbf1466ce54e0"}], "replacing": [], "cryptoCode": "BTC", "trackedSource": "DERIVATIONSCHEME:xpubXXXXXXXXXXXXXXXXXXXX", "transactionData": {"height": null, "blockId": null, "timestamp": 1692603993, "replacedBy": null, "transaction": "02000000000101e3934c150e3b249fe1a7b80f319197e49eae4610826365a6a20b68e8154ea86f01000000000100008002836e0400000000001600143999e3909c4c7200e2fe7552734fbf1466ce54e02d47250000000000160014ad8cdfd59db9321e1f95d8477ae1eadcee347f6902473044022069d9bf1fd404a88e557d4801a3e2ebd369cc4d275e2290740e5eba82a640650102207d000ed6891d7ec50fdade22aaa64301720761fee8dddff6ab9ce7fe2c4c41fe012103ef68e0004ec17595fd920aaee7872e01d3a6f0f55842d0fe58066b00484b89d100000000", "confirmations": 0, "transactionHash": "85b52cecc22077c6a02cf7d45c0452d6f5333b1dbdc02e5ddcdf591da0c19f90"}, "derivationStrategy": "xpubXXXXXXXXXXXXXXXXXXXX"}, "type": "newtransaction"}
```

I expected the `replacing` property of event `531247` to include `96e08ce4b137fd8025fcc7cbc529444e90deb094b65988d8311daed257a3663c`, but it does not.

Here's a screenshot from mempool.space showing that it recognized it as a replacement (I screenshot it because mempool dot space doesn't keep the replacement data forever): 
![Screenshot from 2023-08-21 14-14-57](https://github.com/dgarage/NBXplorer/assets/1530071/a51f79a7-7891-4e2d-87d6-674460e6d250)

Also, decoding the two transactions clearly shows that they spend the same output:

```
$ bcli decoderawtransaction 02000000000101e3934c150e3b249fe1a7b80f319197e49eae4610826365a6a20b68e8154ea86f01000000000000008002836e040000000
0001600143999e3909c4c7200e2fe7552734fbf1466ce54e0514825000000000016001439e9d8d06cc61f11ccff05c8a1afa172e44e6e4d0248304502210081e289c7ddcee3b3e72b6b41597e58d
cfd72c780598aee34d93b927683989f3502203c9f8e4628eb34b47317ddabc5cf50801ec2fe999e02aa2e580457915206373f012103ef68e0004ec17595fd920aaee7872e01d3a6f0f55842d0fe5
8066b00484b89d100000000                                                                                                                                     
{                                                                                                                                                           
  "txid": "96e08ce4b137fd8025fcc7cbc529444e90deb094b65988d8311daed257a3663c",                                                                               
  "hash": "dab8e14be9264eaf814ad998152e28d041ab6eac81f959cc2858218103007e94",                                                                               
  "version": 2,                                                                                                                                             
  "size": 223,                                                                                                                                              
  "vsize": 141,                                                                                                                                             
  "weight": 562,                                                                                                                                            
  "locktime": 0,                                                                                                                                            
  "vin": [                                                                                                                                                  
    {                                                                                                                                                       
      "txid": "6fa84e15e8680ba2a66563821046ae9ee49791310fb8a7e19f243b0e154c93e3",                                                                           
      "vout": 1,                                                                                                                                            
      "scriptSig": {                                                                                                                                        
        "asm": "",                                                                                                                                          
        "hex": ""                                                                                                                                           
      },                                                                                                                                                    
      "txinwitness": [                                                                                                                                      
        "304502210081e289c7ddcee3b3e72b6b41597e58dcfd72c780598aee34d93b927683989f3502203c9f8e4628eb34b47317ddabc5cf50801ec2fe999e02aa2e580457915206373f01", 
        "03ef68e0004ec17595fd920aaee7872e01d3a6f0f55842d0fe58066b00484b89d1"                                                                                
      ],                                                                                                                                                    
      "sequence": 2147483648                                                                                                                                
    }                                                                                                                                                       
  ],                                                                                                                                                        
  "vout": [                                                                                                                                                 
    {                                                                                                                                                       
      "value": 0.00290435,                                                                                                                                  
      "n": 0,                                                                                                                                               
      "scriptPubKey": {                                                                                                                                     
        "asm": "0 3999e3909c4c7200e2fe7552734fbf1466ce54e0",                                                                                                
        "desc": "addr(bc1q8xv78yyuf3eqpch7w4f8xnalz3nvu48qzn5gzy)#are7knuz",                                                                                
        "hex": "00143999e3909c4c7200e2fe7552734fbf1466ce54e0",                                                                                              
        "address": "bc1q8xv78yyuf3eqpch7w4f8xnalz3nvu48qzn5gzy",
        "type": "witness_v0_keyhash"
      }
    },
    {
      "value": 0.02443345,
      "n": 1,
      "scriptPubKey": {
        "asm": "0 39e9d8d06cc61f11ccff05c8a1afa172e44e6e4d",
        "desc": "addr(bc1q885a35rvcc03rn8lqhy2rtapwtjyumjd9h6twf)#ejlakcfe",
        "hex": "001439e9d8d06cc61f11ccff05c8a1afa172e44e6e4d",
        "address": "bc1q885a35rvcc03rn8lqhy2rtapwtjyumjd9h6twf",
        "type": "witness_v0_keyhash"
      }
    }
  ]
}

$ bcli decoderawtransaction 02000000000101e3934c150e3b249fe1a7b80f319197e49eae4610826365a6a20b68e8154ea86f01000000000100008002836e0400000000001600143999e3909c4c7200e2fe7552734fbf1466ce54e02d47250000000000160014ad8cdfd59db9321e1f95d8477ae1eadcee347f6902473044022069d9bf1fd404a88e557d4801a3e2ebd369cc4d275e2290740e5eba82a640650102207d000ed6891d7ec50fdade22aaa64301720761fee8dddff6ab9ce7fe2c4c41fe012103ef68e0004ec17595fd920aaee7872e01d3a6f0f55842d0fe58066b00484b89d100000000
{
  "txid": "85b52cecc22077c6a02cf7d45c0452d6f5333b1dbdc02e5ddcdf591da0c19f90",
  "hash": "8f48be08d0d2d7769fb79602a0c2bdf5d87f15a7ed041b95ea2809b9595e156c",
  "version": 2,
  "size": 222,
  "vsize": 141,
  "weight": 561,
  "locktime": 0,
  "vin": [
    {
      "txid": "6fa84e15e8680ba2a66563821046ae9ee49791310fb8a7e19f243b0e154c93e3",
      "vout": 1,
      "scriptSig": {
        "asm": "",
        "hex": ""
      },
      "txinwitness": [
        "3044022069d9bf1fd404a88e557d4801a3e2ebd369cc4d275e2290740e5eba82a640650102207d000ed6891d7ec50fdade22aaa64301720761fee8dddff6ab9ce7fe2c4c41fe01",
        "03ef68e0004ec17595fd920aaee7872e01d3a6f0f55842d0fe58066b00484b89d1"
      ],
      "sequence": 2147483649
    }
  ],
  "vout": [
    {
      "value": 0.00290435,
      "n": 0,
      "scriptPubKey": {
        "asm": "0 3999e3909c4c7200e2fe7552734fbf1466ce54e0",
        "desc": "addr(bc1q8xv78yyuf3eqpch7w4f8xnalz3nvu48qzn5gzy)#are7knuz",
        "hex": "00143999e3909c4c7200e2fe7552734fbf1466ce54e0",
        "address": "bc1q8xv78yyuf3eqpch7w4f8xnalz3nvu48qzn5gzy",
        "type": "witness_v0_keyhash"
      }
    },
    {
      "value": 0.02443053,
      "n": 1,
      "scriptPubKey": {
        "asm": "0 ad8cdfd59db9321e1f95d8477ae1eadcee347f69",
        "desc": "addr(bc1q4kxdl4vahyepu8u4mprh4c02mnhrglmfekskge)#ml6975pv",
        "hex": "0014ad8cdfd59db9321e1f95d8477ae1eadcee347f69",
        "address": "bc1q4kxdl4vahyepu8u4mprh4c02mnhrglmfekskge",
        "type": "witness_v0_keyhash"
      }
    }
  ]
}
```

**Comment by NicolasDorier** [comment]
@kallerosenbaum the `vin` (`6fa84e15e8680ba2a66563821046ae9ee49791310fb8a7e19f243b0e154c93e3`), does it belongs to a wallet tracked by nbx?

**Comment by NicolasDorier** [comment]
Does the replaced transaction eventually get marked as replaced and remove from mempool in the DB?

NBX normally try to rebroadcast transactions, and should detect it's not possible anymore and update it eventually. (That said, it should be done sooner as you expect)

**Comment by kallerosenbaum** [comment]
> the vin (6fa84e15e8680ba2a66563821046ae9ee49791310fb8a7e19f243b0e154c93e3), does it belongs to a wallet tracked by nbx?
@NicolasDorier No, but weirdly, another output (0) of the same tx is ours (it seems). The vin in the example above spends output 1. Is my understanding correct?

```
SELECT t.*
FROM public.outs t
WHERE tx_id in ('6fa84e15e8680ba2a66563821046ae9ee49791310fb8a7e19f243b0e154c93e3', '0617f1d0b5c3795cbe1b99a3c39eb0d31d7e7174d0a6b4d751d5aef143ea07a3')
```
| code | tx\_id | idx | script | value | asset\_id | input\_tx\_id | input\_idx | input\_mempool | immature | blk\_id | blk\_idx | blk\_height | mempool | replaced\_by | seen\_at |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| BTC | 6fa84e15e8680ba2a66563821046ae9ee49791310fb8a7e19f243b0e154c93e3 | 0 | 0014281516bb775cece999ea92123a8894d0cf1e31c8 | 287177 |  | a8f8bff16b408af7d289f3dd219ec042b62d36c20ddc635f704367d23c9feafa | 996 | false | false | 000000000000000000002e44821515adcb90d2ea93086ce6c2e268daf078787f | 0 | 804067 | false | null | 2023-08-20 13:33:55.861082 +00:00 |

But as you can see, this is not the case for the input tx `0617f1d0b5c3795cbe1b99a3c39eb0d31d7e7174d0a6b4d751d5aef143ea07a3` from the second example, where we don't own any output.

**Comment by NicolasDorier** [comment]
@kallerosenbaum your understanding is correct. Unsure if this is related to the bug yet.

**Comment by NicolasDorier** [comment]
So we get the following situation.

You own: `A1`, but `A0` get double spent and it isn't detected properly.

So this might be normal. As `A1` doesn't belong to the wallet, NBX doesn't detect it because it doesn't try to remember all the inputs of the mempool.

What isn't normal, is that your double spent txs doesn't get its flag back to `mempool=f`.
NBX tries to rebroadcast every 5 minutes, and normally, for the double spent transaction, it detect it and set the flag to false.

I wonder which node/nbx version you are running.

**Comment by kallerosenbaum** [comment]
> NBX normally try to rebroadcast transactions, and should detect it's not possible anymore and update it eventually. (That said, it should be done sooner as you expect)

A few minutes after a replacement wasn't detected as a replacement, I see this in the log for the replaced transaction `4c766e5bfd6214d563ce59984b4625507c2be083348a4d0921baac0bc02de7a9`:

```
[40m[32minfo[39m[22m[49m: NBXplorer.Broadcaster.BTC: Unknown exception when broadcasting 4c766e5bfd6214d563ce59984b4625507c2be083348a4d0921baac0bc02de7a9 (insufficient fee, rejecting replacement 4c766e5bfd6214d563ce59984b4625507c2be083348a4d0921baac0bc02de7a9; new feerate 0.00015921 BTC/kvB <= old feerate 0.00063687 BTC/kvB)
```

This is natural as the replacement tx that pays a higher fee is already in the mempool. But this might be a clue as to why the replaced tx sometimes lingers so long in the database.

**Comment by NicolasDorier** [comment]
This seems weird to me, I have the following code just after:
```sql
else if (result.MissingInput || result.UnknownError)
{
	await conn.ExecuteAsync("UPDATE txs SET mempool='f' WHERE code=@code AND tx_id=@tx_id AND mempool IS TRUE", new { code = tx.Network.CryptoCode, tx_id = tx.Id.ToString() });
}
```

**Comment by NicolasDorier** [comment]
I just pushed ebec46553ca81908531c9518a3555921bbd9ec99

I don't think it fixes the issue, as `Unknown exception when broadcasting` should have set the `mempool='f'` anyway. However, it won't create an entry next time, and it will make this TX as replaced rather as only dropped out of mempool.

**Comment by kallerosenbaum** [comment]
> Does the replaced transaction eventually get marked as replaced and remove from mempool in the DB?

Yes, but sometimes after multiple days as seen in the example above.

**Comment by kallerosenbaum** [comment]
I'm beginning to thing that the "mempool=t" issue is a red herring. I can't tell for sure that there were any instances beyond the one in https://github.com/dgarage/NBXplorer/issues/421#issuecomment-1785255320 and that issue had a very unusual quirk: https://github.com/dgarage/NBXplorer/issues/421#issuecomment-1786612321

**Comment by NicolasDorier** [comment]
Reopening this one, as there is probably two issue, one is why the transaction replacing the other one had empty `replacing`.

The other was, why, despite of it, it wasn't detected as double spent later with the periodic task. (which I just fixed)

**Comment by NicolasDorier** [comment]
However, I believe that my previous fix **will** work.

https://github.com/dgarage/NBXplorer/pull/435 is also properly detecting `rejecting replacement...` messages as a double spend.
Before it was considering as dropping from mempool rather than double spend.

When it drop from mempool, we also update the table `spent_outs` that we use for double spend detection. (note that we update the `spent_outs` table for conflict detection only once per block)

I believe the following events happened:
* Transaction B arrives
* Transaction B' get broadcasted but not yet processed by NBX
* Period task mempool checker mark B as dropping out of mempool
* A new block arrives, removing double spend detection for B in `spent_outs`
* Transaction B' get processed by NBX, double spend is not detected.

Now, the third steps isn't considering B dropping out of mempool, but rather just replaced, as such its inputs aren't removed from `spent_outs`, and the double spend detection should work.

**Comment by NicolasDorier** [comment]
I am closing this comment until you test over #435 on your system. If you still have issue, let's reopen this.

**Comment by kallerosenbaum** [comment]
Given the scenario you described in https://github.com/dgarage/NBXplorer/issues/421#issuecomment-1809323169, It's absolutely plausible that it solves this issue. I'll test this when released in a new version (2.3.67?).

There's just one thing that keeps me hesitant whether this fixes the issue: The scenario you described is a timing issue, where the mempool checker notices the tx is gone before nbx had received `B'`. In this case we should see somewhat different behavior on our two nbx instances (which are also connected to separate nodes), but the two instances exhibit the exact same behavior for a replacement.
